gui/blueprint.lua

```diff
@@ -439,36 +449,14 @@ function BlueprintUI:onRenderBody()
     local start_pos = self.subviews.startpos_panel.start_pos
     if not self.mark and not start_pos then return end
 
-    local vp = self:getViewport()
-    local dc = gui.Painter.new(self.df_layout.map)
-
-    local cursor = df.global.cursor
-    local highlight_bound = self.saved_cursor or cursor
-    local mark = self.mark or highlight_bound
-
-    -- mark start_pos (will get ignored if it's offscreen)
-    if start_pos then
-        local start_tile = vp:tileToScreen(start_pos)
-        dc:map(true):seek(start_tile.x, start_tile.y):
-                pen(COLOR_BLUE, COLOR_BLACK):char('X'):map(false)
+    local function get_overlay_char(pos, is_cursor)
+        -- always render start_pos tile, even if it would overwrite the cursor
+        if same_xy(start_pos, pos) then return 'X', COLOR_BLUE end
+        if is_cursor then return nil end
+        return 'X'
     end
 
-    -- clip scanning range to viewport for performance. offscreen writes will
-    -- get ignored, but it can be slow to scan over large offscreen regions.
-    local _,y_start,y_end = min_to_max(mark.y, highlight_bound.y, vp.y1, vp.y2)
-    local _,x_start,x_end = min_to_max(mark.x, highlight_bound.x, vp.x1, vp.x2)
-    for y=y_start,y_end do
-        for x=x_start,x_end do
-            local pos = xyz2pos(x, y, cursor.z)
-            -- don't overwrite the cursor or start_tile so the user can still
-            -- see them
-            if not same_xy(cursor, pos) and not same_xy(start_pos, pos) then
-                local stile = vp:tileToScreen(pos)
-                dc:map(true):seek(stile.x, stile.y):
-                        pen(COLOR_GREEN, COLOR_BLACK):char('X'):map(false)
-            end
-        end
-    end
+    self:renderMapOverlay(get_overlay_char, self:get_bounds())
```

Comment:
Out of curiosity, did you check if there was any performance impact from switching to this? (In any script, really)

Reply:
I tested every script I modified, and I didn't see any change in performance. F.ex. devel/block-borders uses all the new library code and behaves just the same as before.

Reply:
Except, of course, that it is now more flexible about what screen is active when the script is run